### PR TITLE
Replace `Result<Option<T>, SelectionError>` with `SelectionResult<T>`

### PR DESCRIPTION
**Pull request by compiler-errors**
Not sure if this is a reasonable change or overkill. Also probably needs perf, since this might just be a perf regression due to the hotness of selection.

r? rust-lang/types

### CODE REVIEW

**Comment by lcnr** [comment]
r=me if perf is good

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 1]
```diff
@@ -301,33 +301,30 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
...
                 assert!(self.query_mode == TraitQueryMode::Canonical);
-                return Err(SelectionError::Overflow(OverflowError::Canonical));
-            }
-            Err(SelectionError::Ambiguous(_)) => {
-                return Ok(None);
+                return SelectionResult::Error(SelectionError::Overflow(OverflowError::Canonical));
             }
-            Err(e) => {
-                return Err(e);
+            SelectionResult::Error(SelectionError::Ambiguous(_)) | SelectionResult::Ambiguous => {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:310]: why does `SelectionError::Ambiguous` exist? That's really weird.

It seems like the only place where that's emitted is `fn candidate_from_obligation_on_cache` which should be able to just return `Ok(None)`/ `SelectionResult::Ambiguous`?

**Comment by compiler-errors** [comment]
wow, that sucks.

File: compiler/rustc_trait_selection/src/traits/select/mod.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:310]: It also returns the set of impl candidates that it was not able to winnow down, for error reporting purposes. We could probably return `SelectionResult::Ambiguous`, but then we'd need some other way of asking selcx what impl candidates we should display as possibly applying.

Comment (lcnr) [reply compiler/rustc_trait_selection/src/traits/select/mod.rs:310]: > but then we'd need some other way of asking selcx what impl candidates we should display as possibly applying.

yes, i think that's what we should do here. HAven't looked into how difficult this is, but it definitely feels feasible to me

**Comment by compiler-errors** [comment]
Testing out perf implications of actually doing the the opposite, and removing that extra `Option<T>` by always returning `SelectionError::Ambiguous` on error. That might not be desired, since it does mean we're a bit more fragile -- since we usually need to handle `Err(SelectionError::Ambiguous(_))` case specifically.

**Comment by compiler-errors** [comment]
eh, i don't really care for this change actually, especially lacking a dramatic behavior improvement.

**Pull request closed without merging**